src/file/compression.rs

```diff
@@ -0,0 +1,142 @@
+use anyhow::{Context, Result};
+use filemagic::Magic;
+use log::debug;
+use std::env;
+use std::fs::File;
+use std::path::PathBuf;
+use std::str::FromStr;
+use strum::IntoEnumIterator;
+use strum_macros::EnumIter;
+
+#[derive(Clone, Debug, EnumIter)]
+#[allow(non_camel_case_types)]
+pub enum Compression {
+    xz,
+    bzip2,
+    gzip,
+}
+
+impl FromStr for Compression {
+    type Err = anyhow::Error;
+
+    fn from_str(input: &str) -> Result<Compression, Self::Err> {
+        match input {
+            "xz" => Ok(Compression::xz),
+            "bzip2" => Ok(Compression::bzip2),
+            "gzip" => Ok(Compression::gzip),
+            _ => anyhow::bail!("unknown compression: use either xz, bzip2 or gzip"),
+        }
+    }
+}
+
+impl Compression {
+    fn compress(
+        &self,
+        source: &mut std::fs::File,
+        destination: &mut std::fs::File,
+    ) -> std::io::Result<u64> {
+        let mut enc: Box<dyn std::io::Write> = match &self {
+            Compression::bzip2 => Box::new(bzip2::write::BzEncoder::new(
+                destination,
+                bzip2::Compression::best(),
+            )),
+            Compression::gzip => Box::new(flate2::write::GzEncoder::new(
+                destination,
+                flate2::Compression::best(),
+            )),
+            Compression::xz => {
+                let range = 0..9;
+                let level = env::var("XZ_COMPRESSION_LEVEL")
+                    .unwrap_or_else(|_| "9".to_string())
+                    .parse()
+                    .unwrap_or(9);
+
+                let level = if range.contains(&level) { level } else { 9 };
```

Comment:
nit: maybe consider dropping line 48 for `if (0..9).contains(&level)`. Furthermore, while the result is the same, due to the else case, the range should be inclusive, i.e. `0..=9`.